File: Assistant/lib/utility/convert_helper.dart
import '/common.dart';
import 'dart:async';
import 'dart:typed_data';
import 'dart:ui';
import 'package:decimal/decimal.dart' as lib;
import 'package:stack_trace/stack_trace.dart' as lib;

// ----------------

class ConvertHelper {

  // #region exception

  static String generateExceptionMessage(
    Object      exception,
    StackTrace? stack,
  ) {
    var result = '${exception}';
    if (stack != null) {
      var trace = lib.Trace.from(stack);
      for (var frame in trace.frames) {
        result += '\n@ ${frame.library.selfLet((it) => it.startsWith('package:assistant') ? it.substring('package:'.length) : it)}:${frame.line ?? '?'}:${frame.column ?? '?'} ${frame.member ?? '?'}';
      }
    }
    return result;
  }

  // #endregion

  // #region enumeration

  static String makeEnumerationToStringOfSnakeCase<TValue extends Enum>(
    TValue value,
  ) {
    return ConvertHelper.changeStringFromCamelCaseToSnakeCase(value.name);
  }

  static TValue parseEnumerationFromStringOfSnakeCase<TValue extends Enum>(
    String       text,
    List<TValue> list,
  ) {
    return list.firstWhere((it) => ConvertHelper.makeEnumerationToStringOfSnakeCase(it) == text);
  }

  // #endregion

  // #region boolean

  static String makeBooleanToString(
    Boolean value,
  ) {
    return switch (value) {
      false => 'false',
      true  => 'true',
    };
  }

  static String makeBooleanToStringOfConfirmationCharacter(
    Boolean value,
  ) {
    return switch (value) {
      false => 'n',
      true  => 'y',
    };
  }

  // #endregion

  // #region integer

  static String makeIntegerToString(
    Integer value,
  ) {
    return '${value}';
  }

  // #endregion

  // #region floater

  static String makeFloaterToString(
    Floater value,
  ) {
    var valueDecimal = lib.Decimal.parse(value.toString());
    return '${valueDecimal}${(!valueDecimal.isInteger ? '' : '.0')}';
  }

  // #endregion

  // #region character

  static Boolean isLetter(
    String value,
  ) {
    if (value.length != 1) {
      return false;
    }
    var code = value.codeUnitAt(0) | 0x20;
    return 'a'.codeUnitAt(0) <= code && code <= 'z'.codeUnitAt(0);
  }

  // ----------------

  static Boolean isPathDot(
    String value,
  ) {
    if (value.length != 1) {
      return false;
    }
    return value == '.';
  }

  static Boolean isPathDirectorySeparator(
    String value,
  ) {
    if (value.length != 1) {
      return false;
    }
    return value == '/' || value == '\\';
  }

  static Boolean isPathVolumeSeparator(
    String value,
  ) {
    if (value.length != 1) {
      return false;
    }
    return value == ':';
  }

  // #endregion

  // #region string

  static List<String> splitStringOfCamelCase(
    String source,
  ) {
    return RegExp(r'([A-Z][a-z]*)|([a-z]+)|([0-9]+)')
      .allMatches(source)
      .map((it) => it.group(0)!)
      .toList();
  }

  // ----------------

  static String changeStringFromCamelCaseToSnakeCase(
    String source,
  ) {
    return ConvertHelper.splitStringOfCamelCase(source).join('_').toLowerCase();
  }

  // #endregion

  // #region string list

  static String makeStringListToStringWithLine(
    List<String> value,
  ) {
    return value.join('\n') + (!value.isEmpty && value.last.isEmpty ? '\n' : '');
  }

  static List<String> parseStringListFromStringWithLine(
    String text,
  ) {
    var value = text.split('\n').map((value) => !value.endsWith('\r') ? value : value.substring(0, value.length - 1)).toList();
    if (!value.isEmpty && value.last.isEmpty) {
      value.removeLast();
    }
    return value;
  }

  // #endregion

  // #region image

  static Future<Image> parseImageFromData(
    Uint8List data, {
    Integer?  width = null,
    Integer?  height = null,
    Boolean   isRawRgba = false,
    Boolean   isRawBgra = false,
    Boolean   isPng = false,
  }) async {
    assertTest([isRawRgba, isRawBgra, isPng].where((it) => it).length == 1);
    var value = null as Image?;
    if (isPng) {
      var codec = await instantiateImageCodec(data);
      var frame = await codec.getNextFrame();
      value = frame.image;
    }
    if (isRawRgba) {
      var completer = Completer<Image>();
      decodeImageFromPixels(data, width!, height!, .rgba8888, (image) => completer.complete(image));
      value = await completer.future;
    }
    if (isRawBgra) {
      var completer = Completer<Image>();
      decodeImageFromPixels(data, width!, height!, .bgra8888, (image) => completer.complete(image));
      value = await completer.future;
    }
    return value!;
  }

  // #endregion

}

File: AssistantPlus/Application/Utility/ConvertHelper.cs
#pragma warning disable 0,
// ReSharper disable

using Twinning.AssistantPlus;
using System.Runtime.InteropServices.WindowsRuntime;
using Windows.Graphics.Imaging;
using Windows.UI;
using Microsoft.UI.Xaml.Media.Imaging;
using Windows.ApplicationModel.DataTransfer;

namespace Twinning.AssistantPlus.Utility {

	public static class ConvertHelper {

		#region type

		public static Boolean IsTypeOfTuple(
			Type type
		) {
			if (!type.IsGenericType) {
				return false;
			}
			var genericType = type.GetGenericTypeDefinition();
			return genericType == typeof(Tuple<>)
				|| genericType == typeof(Tuple<,>)
				|| genericType == typeof(Tuple<,,>)
				|| genericType == typeof(Tuple<,,,>)
				|| genericType == typeof(Tuple<,,,,>)
				|| genericType == typeof(Tuple<,,,,,>)
				|| genericType == typeof(Tuple<,,,,,,>)
				|| genericType == typeof(Tuple<,,,,,,,>);
		}

		public static Boolean IsTypeOfValueTuple(
			Type type
		) {
			if (!type.IsGenericType) {
				return false;
			}
			var genericType = type.GetGenericTypeDefinition();
			return genericType == typeof(ValueTuple<>)
				|| genericType == typeof(ValueTuple<,>)
				|| genericType == typeof(ValueTuple<,,>)
				|| genericType == typeof(ValueTuple<,,,>)
				|| genericType == typeof(ValueTuple<,,,,>)
				|| genericType == typeof(ValueTuple<,,,,,>)
				|| genericType == typeof(ValueTuple<,,,,,,>)
				|| genericType == typeof(ValueTuple<,,,,,,,>);
		}

		#endregion

		#region exception

		public static String GenerateExceptionMessage(
			Exception exception
		) {
			var message = $"{exception.Message}";
			if (exception.StackTrace != null) {
				foreach (var frame in exception.StackTrace.Split(Environment.NewLine)) {
					if (!frame.StartsWith("   at ")) {
						continue;
					}
					message += $"\n@ {frame.Substring("   at ".Length)}";
				}
			}
			return message;
		}

		#endregion

		#region enumeration

		public static String MakeEnumerationToStringOfSnakeCase<TValue>(
			TValue value
		)
			where TValue : struct, Enum {
			return ConvertHelper.ChangeStringFromCamelCaseToSnakeCase(Enum.GetName(value).AsNotNull());
		}

		public static TValue ParseEnumerationFromStringOfSnakeCase<TValue>(
			String text
		)
			where TValue : struct, Enum {
			return Enum.GetValues<TValue>().First((it) => ConvertHelper.MakeEnumerationToStringOfSnakeCase(it) == text);
		}

		#endregion

		#region boolean

		public static String MakeBooleanToString(
			Boolean value
		) {
			return value switch {
				false => "false",
				true  => "true",
			};
		}

		public static String MakeBooleanToStringOfConfirmationCharacter(
			Boolean value
		) {
			return value switch {
				false => "n",
				true  => "y",
			};
		}

		public static Floater MakeBooleanToFloaterOfOpacityVisibility(
			Boolean value
		) {
			return value switch {
				false => 0.0,
				true  => 1.0,
			};
		}

		public static Floater MakeBooleanToFloaterOfOpacityEnabled(
			Boolean value
		) {
			return value switch {
				false => 0.5,
				true  => 1.0,
			};
		}

		#endregion

		#region integer

		public static String MakeIntegerToString(
			Integer value,
			Boolean showPositiveSign
		) {
			return value switch {
				0   => "0",
				< 0 => $"{value}",
				> 0 => $"{(!showPositiveSign ? "" : "+")}{value}",
			};
		}

		#endregion

		#region floater

		public static String MakeFloaterToString(
			Floater value,
			Boolean showPositiveSign
		) {
			var valueDecimal = new Decimal(value);
			return value switch {
				0.0   => "0",
				< 0.0 => $"{valueDecimal}",
				> 0.0 => $"{(!showPositiveSign ? "" : "+")}{valueDecimal}",
				_     => throw new (),
			} + (!Decimal.IsInteger(valueDecimal) ? "" : ".0");
		}

		#endregion

		#region character

		public static Boolean IsPathSeparator(
			Character value
		) {
			return value == '/' || value == '\\';
		}

		public static Boolean IsLetter(
			Character value
		) {
			var code = (value | 0x20).CastPrimitive<Character>();
			return 'a' <= code && code <= 'z';
		}

		#endregion

		#region string

		public static List<String> SplitStringOfCamelCase(
			String source
		) {
			return new Regex(@"([A-Z][a-z]*)|([a-z]+)|([0-9]+)", RegexOptions.Compiled)
				.Matches(source)
				.Select((it) => it.Value)
				.ToList();
		}

		// ----------------

		public static String ChangeStringFromCamelCaseToSnakeCase(
			String source
		) {
			return String.Join('_', ConvertHelper.SplitStringOfCamelCase(source)).ToLower();
		}

		public static String ChangeStringFromCamelCaseThenInsertSpace(
			String source
		) {
			return String.Join(' ', ConvertHelper.SplitStringOfCamelCase(source));
		}

		#endregion

		#region string list

		public static String MakeStringListToStringWithLine(
			List<String> value
		) {
			return String.Join('\r', value) + (!value.IsEmpty() && value[^1].IsEmpty() ? "\r" : "");
		}

		public static List<String> ParseStringListFromStringWithLine(
			String text
		) {
			var value = text.Split('\r').ToList();
			if (!value.IsEmpty() && value[^1].IsEmpty()) {
				value.RemoveAt(value.Count - 1);
			}
			return value;
		}

		#endregion

		#region color

		public static Integer MakeColorToInteger(
			Color value
		) {
			return (value.A.CastPrimitive<Integer>() << 24)
				| (value.R.CastPrimitive<Integer>() << 16)
				| (value.G.CastPrimitive<Integer>() << 8)
				| (value.B.CastPrimitive<Integer>() << 0);
		}

		public static Color ParseColorFromInteger(
			Integer integer
		) {
			return Color.FromArgb(
				((integer & 0xFF000000L) >> 24).CastPrimitive<Byte>(),
				((integer & 0x00FF0000L) >> 16).CastPrimitive<Byte>(),
				((integer & 0x0000FF00L) >> 8).CastPrimitive<Byte>(),
				((integer & 0x000000FFL) >> 0).CastPrimitive<Byte>()
			);
		}

		#endregion

		#region bitmap

		public static async Task<WriteableBitmap> ParseBitmapFromBinary(
			Byte[] data
		) {
			using var stream = data.AsBuffer().AsStream().AsRandomAccessStream();
			var decoder = await BitmapDecoder.CreateAsync(stream);
			var image = new WriteableBitmap(decoder.PixelWidth.CastPrimitive<Size>(), decoder.PixelHeight.CastPrimitive<Size>());
			stream.Seek(0);
			await image.SetSourceAsync(stream);
			return image;
		}

		#endregion

		#region data view

		public static async Task<List<StoragePath>> DataViewGetStoragePath(
			DataPackageView view
		) {
			return (await view.GetStorageItemsAsync())
				.Select((it) => StorageHelper.GetLongPath(it.Path))
				.ToList();
		}

		#endregion

	}

}
